llvm/lib/Transforms/Utils/BasicBlockUtils.cpp

```diff
@@ -1909,22 +1909,29 @@ static void reconnectPhis(BasicBlock *Out, BasicBlock *GuardBlock,
     auto NewPhi =
         PHINode::Create(Phi->getType(), Incoming.size(),
                         Phi->getName() + ".moved", FirstGuardBlock->begin());
+    bool AllUndef = true;
     for (auto *In : Incoming) {
       Value *V = UndefValue::get(Phi->getType());
       if (In == Out) {
         V = NewPhi;
       } else if (Phi->getBasicBlockIndex(In) != -1) {
         V = Phi->removeIncomingValue(In, false);
+        AllUndef &= isa<UndefValue>(V);
       }
       NewPhi->addIncoming(V, In);
     }
     assert(NewPhi->getNumIncomingValues() == Incoming.size());
+    Value *NewV = NewPhi;
+    if (AllUndef) {
+      NewPhi->eraseFromParent();
+      NewV = UndefValue::get(Phi->getType());
```

Comment:
Should this try to preserve poison?

Reply:
As of right now, this emits undef by default. We likely want to change both UndefValue::get to poison. Should I include that and track undef/poison separately?

Reply:
It probably makes little difference, can switch these to poison as a second step?

Reply:
OK, let's do this first, then we switch to poison.

Reply:
I didn't really understand. Are we agreeing that it's okay to temporarily ignore `UndefValue` that are actually poison in this patch?

Reply:
All poisons are undef. Not all undefs are poison. It's safe, though suboptimal, to replace poison with undef